src/xrpld/app/tx/detail/SetTrust.cpp

```diff
@@ -408,15 +413,35 @@ SetTrust::doApply()
             uFlagsOut &= ~(bHigh ? lsfHighNoRipple : lsfLowNoRipple);
         }
 
+        bool willSetFreeze = false;
         if (bSetFreeze && !bClearFreeze && !sle->isFlag(lsfNoFreeze))
         {
             uFlagsOut |= (bHigh ? lsfHighFreeze : lsfLowFreeze);
+            willSetFreeze = true;
         }
         else if (bClearFreeze && !bSetFreeze)
         {
             uFlagsOut &= ~(bHigh ? lsfHighFreeze : lsfLowFreeze);
         }
 
+        bool const alreadyFrozen =
+            bHigh ? sle->isFlag(lsfHighFreeze) : sle->isFlag(lsfLowFreeze);
+        if (bSetDeepFreeze && !bClearDeepFreeze &&
+            (willSetFreeze || alreadyFrozen))
+        {
+            if (ctx_.view().rules().enabled(featureDeepFreeze))
+                uFlagsOut |= (bHigh ? lsfHighDeepFreeze : lsfLowDeepFreeze);
+            else
+                return temMALFORMED;
+        }
+        else if (bClearDeepFreeze && !bSetDeepFreeze)
+        {
+            if (ctx_.view().rules().enabled(featureDeepFreeze))
+                uFlagsOut &= ~(bHigh ? lsfHighDeepFreeze : lsfLowDeepFreeze);
+            else
+                return temMALFORMED;
+        }
```

Comment:
I disagree for two reasons:
1. It goes against the design behind the division of the three stages of checking:
   1. https://github.com/XRPLF/rippled/blob/develop/src/xrpld/app/tx/applySteps.h#L229-L232
      * The transaction is checked against all possible validity constraints that do not require a ledger.
   2. https://github.com/XRPLF/rippled/blob/develop/src/xrpld/app/tx/applySteps.h#L253-L264
      * The transaction is checked against all possible validity constraints that DO require a ledger.
   3. https://github.com/XRPLF/rippled/blob/develop/src/xrpld/app/tx/applySteps.h#L320-L323
      * Apply a prechecked transaction to an OpenView.
      * In other words, `preflight` and `preclaim` should check as many things as possible before the transaction ever gets to `doApply`, so that it can avoid checks and _just_ make the necessary changes. There are some exceptions, but they should be limited to complicated transactions where the checks are very difficult to avoid, e.g. stuff using the payment engine, but this is not a complicated transactoin.
2. This solution missed one of the checks. It still ignores attempts to set the freeze flag on accounts with `lsfNoFreeze` set.
```
        // New behavior with this amendment: If lsfNoFreeze is set, any attempt
        // to set a freeze will fail and claim a fee instead of just being
        // ignored
        if ((willFreeze || willDeepFreeze) && sle->isFlag(lsfNoFreeze))
            return tecNO_PERMISSION;
```

> I think putting that into preclaim would make it more complicated actually.

This part of `doApply` becomes a bit simpler when the checks are moved to `preclaim`
```
       if (bSetFreeze && !bClearFreeze)
        {
            uFlagsOut |= (bHigh ? lsfHighFreeze : lsfLowFreeze);
        }
        else if (bClearFreeze && !bSetFreeze)
        {
            uFlagsOut &= ~(bHigh ? lsfHighFreeze : lsfLowFreeze);
        }

        if (bSetDeepFreeze && !bClearDeepFreeze)
        {
            uFlagsOut |= (bHigh ? lsfHighDeepFreeze : lsfLowDeepFreeze);
        }
        else if (bClearDeepFreeze && !bSetDeepFreeze)
        {
            uFlagsOut &= ~(bHigh ? lsfHighDeepFreeze : lsfLowDeepFreeze);
        }
```

Reply:
Could you elaborate on point 2? I'm pretty sure lsfNoFreeze is checked for both freeze and deep freeze

Reply:
> Could you elaborate on point 2? I'm pretty sure lsfNoFreeze is checked for both freeze and deep freeze

`lsfNoFreeze` is checked, but if it's set, [this code](https://github.com/XRPLF/rippled/pull/5187#discussion_r1846987373) silently ignores it. We want new behavior with the amendment to also return `tecNO_PERMISSION`.
```
        // New behavior with this amendment: If lsfNoFreeze is set, any attempt
        // to set a freeze will fail and claim a fee instead of just being
        // ignored
```

Reply:
Yes, it's been fixed.

The old behavior remains as is